common/middleware/src/line_rules.rs

```diff
@@ -0,0 +1,293 @@
+use crate::{Middleware, Status};
+use http::types::body::LineBufferMut;
+use regex::bytes::Regex;
+use std::cmp;
+use thiserror::Error;
+
+static REDACT_BYTES: &[u8] = "[REDACTED]".as_bytes();
+
+pub struct LineRules {
+    exclusion: Vec<Regex>,
+    inclusion: Vec<Regex>,
+    redact: Vec<Regex>,
+}
+
+#[derive(Clone, Debug, Error)]
+pub enum LineRulesError {
+    #[error(transparent)]
+    RegexError(regex::Error),
+}
+
+impl LineRules {
+    pub fn new(
+        exclusion: &[String],
+        inclusion: &[String],
+        redact: &[String],
+    ) -> Result<LineRules, LineRulesError> {
+        Ok(LineRules {
+            exclusion: map_to_regex(exclusion)?,
+            inclusion: map_to_regex(inclusion)?,
+            redact: map_to_regex(redact)?,
+        })
+    }
+
+    /// Applies inclusion and exclusion rules and replaces the redacted values.
+    fn process_line<'a>(
+        &self,
+        line: &'a mut dyn LineBufferMut,
+    ) -> Status<&'a mut dyn LineBufferMut> {
+        let value = line.get_line_buffer().unwrap();
+
+        // If it doesn't match any inclusion rule -> skip
+        if !self.inclusion.is_empty() && !self.inclusion.iter().any(|r| r.is_match(value)) {
```

Comment:
Since we don't use the match locations anywhere the inclusion and exclusion rules could be turned into RegexSets, which would make the match marginally quicker since it can match in a single pass and the Regexset can optimise it's representation across all the Regexs at once

Reply:
sure